tests/e2e.py

```diff
@@ -0,0 +1,117 @@
+        "execution_started_at": "2024-10-26T14:15:16.400388Z",
+        "execution_ended_at": "2024-10-26T14:15:16.5454Z",
+        "result": {
+            "rows": [
+                {
+                    "block_date": "2024-09-28",
+                    "block_number": 20849352,
+                    "block_time": "2024-09-28 13:12:11.000 UTC",
+                    "hash": "0x5f0b3f5d3f15bf9943b1b6c77f69",
+                    "success": True,
+                    "type": "DynamicFee",
+                }
+            ],
+            "metadata": {
+                "column_names": [
+                    "block_time",
+                    "block_number",
+                    "success",
+                    "hash",
+                    "type",
+                    "block_date",
+                ],
+                "column_types": [
+                    "timestamp with time zone",
+                    "bigint",
+                    "boolean",
+                    "varbinary",
+                    "varchar",
+                    "date",
+                ],
+                "row_count": 1,
+                "result_set_bytes": 97,
+                "total_row_count": 1,
+                "total_result_set_bytes": 97,
+                "datapoint_count": 6,
+                "pending_time_millis": 352,
+                "execution_time_millis": 145,
+            },
+        },
+    }
+)
+
+
+class TestEndToEnd(unittest.TestCase):
+    def test_dune_results_to_db(self):
+
+        engine = create_engine(DB_URL)
+        df, types = dune_result_to_df(SAMPLE_DUNE_RESULTS.result)
+
+        expected = DataFrame(
+            {
+                "block_date": ["2024-09-28"],
+                "block_number": [20849352],
+                "block_time": ["2024-09-28 13:12:11.000 UTC"],
+                "hash": [b"_\x0b?]?\x15\xbf\x99C\xb1\xb6\xc7\x7fi"],
+                "success": [True],
+                "type": ["DynamicFee"],
+            }
+        )
+        self.assertIsNone(
+            pandas.testing.assert_frame_equal(df, expected, check_dtype=True)
+        )
+        self.assertEqual(
+            types,
+            {
+                "block_date": DATE,
+                "block_number": BIGINT,
+                "block_time": TIMESTAMP,
+                "hash": BYTEA,
+                "success": BOOLEAN,
+                "type": VARCHAR,
+            },
+        )
+
+        save_to_postgres(engine, "test_table", df, types)
+
+        self.assertEqual(
+            query_pg(engine, "select * from test_table"),
+            [
+                {
+                    "block_date": datetime.date(2024, 9, 28),
+                    "block_number": 20849352,
+                    "block_time": datetime.datetime(2024, 9, 28, 13, 12, 11),
+                    "hash": "0x5f0b3f5d3f15bf9943b1b6c77f69",
+                    "success": True,
+                    "type": "DynamicFee",
+                }
+            ],
+        )
+
+
+if __name__ == "__main__":
```

Comment:
nitpick: this block is unnecessary because we're running pytest over the tests dir, and most IDEs will let you run the test class by itself anyway.

Reply:
Interesting. I'll try without and if it works... I'll remove it. I usually run my tests from cli.

Reply:
`python -m pytest tests/` also runs that test module just fine without the `__name__ == "__main__":` part.

Reply:
It worked out just fine.